internal/dot/graph.go

```diff
@@ -60,12 +71,27 @@ type Group struct {
 	Type    reflect.Type
 	Name    string
 	Results []*Result
+	State   State
+
+	// ResultMap maps constructor ID to the grouped result it produces so we
+	// can get the result that failed given constructor ID.
+	ResultMap map[uintptr]*Result
 }
 
 // Graph is the DOT-format graph in a Container.
 type Graph struct {
-	Ctors  []*Ctor
-	Groups map[groupKey]*Group
+	Ctors   []*Ctor
+	ctorMap map[uintptr]*Ctor
+	Groups  map[groupKey]*Group
+
+	// Pof is a list of the point of failures. They are the root causes of
+	// failed invokes and can be either missing types (not provided) or error
+	// types (error providing).
+	Pof []*Result
+
+	// Failed is the list of nodes that failed to build due to missing/failed
+	// dependencies.
+	Failed []*Result
```

Comment:
Is there overlap with root causes or are they separate?

Reply:
No, they are disjoint sets.